### DISCUSSION TITLE
TFLite mismatch: Didn't find op for builtin opcode 'CONV_2D' version '3'

### DISCUSSION THREAD

**Discussion by ksaye** [pschatzmann/arduino-audio-tools#657] [Q&A] [answered]
Following https://github.com/pschatzmann/arduino-audio-tools/wiki/TensorFlow-Lite---MicroSpeech, I was able to generate the model file: https://kevinsayazstorage.blob.core.windows.net/public/model.h.

Using the sample stream-i2s-tf works great with the included model.h.

When I add my model and make the following change (I changed the WANTED_WORDS = "marvin,stop,off,on,go,sheila") in the notebook:
`const char* kCategoryLabels[8] = {"silence","unknown","marvin","stop","off","on","go","sheila"};`

I get an error:

      Didn't find op for builtin opcode 'CONV_2D' version '3'. An older version of this builtin might be supported. Are you using an old TFLite binary with a newer model?
      
      Failed to get registration from op code CONV_2D
       
      [E] TfLiteAudioStream.h : 857 - AllocateTensors() failed
      Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.
      Core 1 register dump:
      PC      : 0x4000c2e4  PS      : 0x00060b30  A0      : 0x800d31d7  A1      : 0x3ffb1eb0  
      A2      : 0x00000000  A3      : 0x3ffd03a0  A4      : 0x000007a8  A5      : 0x00000000  
      A6      : 0x00000000  A7      : 0x00000000  A8      : 0x00000000  A9      : 0x000000b1  
      A10     : 0x0001cd4d  A11     : 0x00000028  A12     : 0x3ffbc018  A13     : 0x00000016  
      A14     : 0x3ffbc068  A15     : 0x00000000  SAR     : 0x0000001a  EXCCAUSE: 0x0000001d  
      EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000079  
      
      ELF file SHA256: 0000000000000000
      
      Backtrace: 0x4000c2e4:0x3ffb1eb0 0x400d31d4:0x3ffb1ec0 0x400d2d8e:0x3ffb1ef0 0x400d2e5d:0x3ffb1f10 0x400d4b82:0x3ffb1f30 0x400d4d1b:0x3ffb1f50 0x400d4d86:0x3ffb1f90 0x40105b79:0x3ffb1fb0 0x4008649e:0x3ffb1fd0

      Rebooting...

Is this a known issue?

**Comment by pschatzmann** [comment]
My first guess would be that it is running out of RAM. You would need to analyze the stacktrace to confirm this.
If you have PSRAM you could try to activate the use of PSRAM 
https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/

**Comment by ksaye** [comment]
I just recompiled the model, this time with fewer words.  The sample model.h you include [char array] is of size: 18715.  My model is of size 14752, which is smaller and I get the same error.

Are you sure there has not been an update in the model?  I see a cell in the notebook runs: `git clone -q --depth 1 https://github.com/tensorflow/tensorflow` and I am wondering if something in the tenforflow repo has changed.

**Reply by ksaye** [reply]
I also get the same error with: https://github.com/pschatzmann/tflite-micro-arduino-examples/blob/main/examples/micro_speech/micro_speech.ino, once I replace the model.cc and adjust the kCategoryLabels.

    setup - begin
    Didn't find op for builtin opcode 'CONV_2D' version '3'. An older version of this builtin might be supported. Are you using an old TFLite binary with a newer model?
    
    Failed to get registration from op code CONV_2D
     
    AllocateTensors() failed
    Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
    Core 1 register dump:
    PC      : 0x400d32c3  PS      : 0x00060130  A0      : 0x800d3524  A1      : 0x3ffb1f50  
    A2      : 0x00000000  A3      : 0x3ffbdbcc  A4      : 0x00000000  A5      : 0x00000028  
    A6      : 0x3ffb1f88  A7      : 0x3ffc07b8  A8      : 0x00000002  A9      : 0x3ffb1f40  
    A10     : 0x3ffba4e4  A11     : 0x00000000  A12     : 0x00000029  A13     : 0x0000efc0  
    A14     : 0x00000008  A15     : 0x3f400e94  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
    EXCVADDR: 0x00000000  LBEG    : 0x40088811  LEND    : 0x40088821  LCOUNT  : 0xfffffff9  
    
    ELF file SHA256: 0000000000000000
    
    Backtrace: 0x400d32c3:0x3ffb1f50 0x400d3521:0x3ffb1f80 0x400df78d:0x3ffb1fb0 0x4008a33e:0x3ffb1fd0
    
    Rebooting...

**Reply by ksaye** [reply]
@pschatzmann 

I have spent days trying to resolve this.  Using your model.h it works.  When I compile my own, I get the error below.  I have tried a different tag version of tensorflow, nothing works.

I still get:

    16:04:48.365 -> Didn't find op for builtin opcode 'CONV_2D' version '3'. An older version of this builtin might be supported. Are you using an old TFLite binary with a newer model?
    16:04:48.397 -> 
    16:04:48.397 -> Failed to get registration from op code CONV_2D
    16:04:48.397 ->  
    16:04:48.397 -> [E] TfLiteAudioStream.h : 857 - AllocateTensors() failed
    16:04:48.397 -> Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.
    16:04:48.429 -> Core 1 register dump:
    16:04:48.429 -> PC      : 0x4000c2e4  PS      : 0x00060b30  A0      : 0x800d31d7  A1      : 0x3ffb1eb0  
    16:04:48.429 -> A2      : 0x00000000  A3      : 0x3ffc7c58  A4      : 0x000007a8  A5      : 0x00000000  
    16:04:48.429 -> A6      : 0x00000000  A7      : 0x00000000  A8      : 0x00000000  A9      : 0x000000b3  
    16:04:48.429 -> A10     : 0x0001d24d  A11     : 0x00000028  A12     : 0x3ffbab90  A13     : 0x00000016  
    16:04:48.472 -> A14     : 0x3ffbabe0  A15     : 0x00000000  SAR     : 0x0000001a  EXCCAUSE: 0x0000001d  
    16:04:48.472 -> EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000079  
    16:04:48.472 -> 
    16:04:48.472 -> ELF file SHA256: 0000000000000000
    16:04:48.472 -> 
    16:04:48.472 -> Backtrace: 0x4000c2e4:0x3ffb1eb0 0x400d31d4:0x3ffb1ec0 0x400d2d8e:0x3ffb1ef0 0x400d2e5d:0x3ffb1f10 0x400d4b82:0x3ffb1f30 0x400d4d1b:0x3ffb1f50 0x400d4d86:0x3ffb1f90 0x40105b79:0x3ffb1fb0 0x4008649e:0x3ffb1fd0
    16:04:48.502 -> 
    16:04:48.502 -> Rebooting...

When I try to run: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-i2s-tf

As you can see from: [fail_train_micro_speech_model.zip](https://github.com/pschatzmann/arduino-audio-tools/files/10893077/fail_train_micro_speech_model.zip), I can train the model but when I copy model.cc it compiles but errors as above.  Can you take a quick look?  [model.zip](https://github.com/pschatzmann/arduino-audio-tools/files/10893091/model.zip)

**Comment by pschatzmann** [comment, marked as answer]
When I created the library it was for version=2.4.0-ALPHA
Did you try to make sure that you used the correspinding version in Jupyter ?

**Reply by ksaye** [reply]
I have tried a few version, but not that version.  Let me git it a whirl.

**Reply by ksaye** [reply]
I tried in cell 5 to use `!git clone -q --depth 1 https://github.com/tensorflow/tensorflow --branch v2.4.0` and `!git clone -q --depth 1 https://github.com/tensorflow/tensorflow --branch v2.3.0`, neither worked.

**Reply by KasunThushara** [reply]
I tried creating my own model with a custom dataset. I tested with TensorFlow versions **2.4.1** and **2.3.1**.
[Train Micro Speech Model Notebook](https://colab.research.google.com/github/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/examples/micro_speech/train/train_micro_speech_model.ipynb#scrollTo=lBj_AyCh1cC0)


* For TensorFlow 2.4.1, I downloaded the corresponding TensorFlow source with:

```bash
!git clone -q --depth 1 https://github.com/tensorflow/tensorflow --branch v2.4
```

* For TensorFlow 2.3.1, I used:

```bash
!git clone -q --depth 1 https://github.com/tensorflow/tensorflow --branch v2.3.1
```

The models were successfully created, but I keep encountering this error in the Arduino code:

```
Didn't find op for builtin opcode 'CONV_2D' version '3'. 
An older version of this builtin might be supported. 
Are you using an old TFLite binary with a newer model?
Failed to get registration from op code CONV_2D
```




Do you have any updates on how to resolve this error?

**Reply by pschatzmann** [reply]
Did you try with the tflite that is part of the esp32 core?

**Reply by KasunThushara** [reply]
for compile the model I used 1.5 TF . the model was compiled and it works with audio-tools
